### PR TITLE
feat(auth-guard): add support for specifying a `string` to redirect to

### LINKED ISSUE #2144: Feature Request: Add queryParams into return url from custom AuthGuardPipe
I'm trying use authGuard to return deep-link for user after success login but it's look like return url from authGuardPipe be encoded, therefore my link break. 

my code: 

```
const redirectUnauthorizedUser = (next: ActivatedRouteSnapshot, ) => map(user => {
  if (!user) {
    return ['/guest/login?returnUrl=${encodeURI(next.pathFromRoot
      .map(v => v.url.map(segment => segment.toString()).join('/'))
      .join('/'))}'];
  } else {
    return true;
  }
});

const routes: Routes = [
  {
    path: 'guest',
    loadChildren: () => import('./component/guest/guest.module').then(module => module.GuestModule),
    canActivate: [AngularFireAuthGuard], data: { authGuardPipe: redirectLoggedInUser }
  },
  {
    path: 'account',
    loadChildren: () => import('./component/account/account.module').then(module => module.AccountModule),
    canActivate: [AngularFireAuthGuard], data: { authGuardPipe: redirectUnauthorizedUser }
  }
];
```

**Request:**
1, Is there anyway to add query param into url that return from custom authGuardPipe.
2, It would be great if i can use RouterState in custom pipe for get queryparams in a proper way, not like the way i did above.

Thank you.

**UPDATED**
bacause `AngularFireAuthGuard implements CanActivate` therefore i can use `routerState` like normal CanActivate. But i still not found solution for question 1.

### LINKED ISSUE #2287: feat(redirectUnauthorizedTo): allow passing query parameters
### Version info

**Angular:**
~9.0.0-rc.7

**Firebase:**
^7.6.1-0

**AngularFire:**
^5.3.0-rc.4

Using the `AngularFireAuthGuard` works, but does not support redirect with query parameters.
When a user isn't authenticated, I'd like to direct them to login, and once they are logged in send them back to where they tried to access.

```ts
const redirectUnauthorizedToLogin = (route: ActivatedRouteSnapshot) => {
  const path = route.pathFromRoot.map(v => v.url.map(segment => segment.toString()).join('/')).join('/');
  const params = String(new URLSearchParams(route.queryParams));

  // const url = '/?' + String(new URLSearchParams({redirect: path + '?' + params}));

  return redirectUnauthorizedTo(['/']);
};
```

But I can't redirect to this url as its not a valid segment, and `redirectUnauthorizedTo` doesn't support query parameters

### PR DESCRIPTION
**Pull request by EdricChan03**
Closes #2287
Closes #2144

### Checklist

   - Issue number for this PR: #2287, #2144
   - Docs included?: no 
   - Test units included?: no
   - In a clean directory, `yarn install`, `yarn test` run successfully? (yes/no; required)

### Description

This PR introduces the ability to specify a `UrlTree` or a `string` to the return value of an `AuthPipe` and includes this feature in the existing `redirect*To` methods.

Previously, as noted in #2287, it was not possible to specify additional extras that could be passed to [`Router#createUrlTree`](https://angular.io/api/router/Router#createUrlTree).

### Code sample

- With a string:
    ```typescript
    export const redirectUnauthorizedToLogin = (afterLoginRedirect: string) => redirectUnauthorizedTo(`/login?redirectTo=${afterLoginRedirect}`);
    // ...
    const routes: Route[] = [
      {
        path: 'items',
        component: ItemsComponent
        ...canActivate(redirectUnauthorizedToLogin('/items'))
      }
    ]
    ```
- With a [`UrlTree`](https://angular.io/api/router/UrlTree) (not currently possible atm):
    ```typescript
    // TODO
    ```

### CODE REVIEW

**Comment by EdricChan03** [comment]
Seems that I did not think through the `UrlTree` portion - how should this be implemented?

File: src/auth-guard/auth-guard.ts [hunk 1]
```diff
@@ -40,21 +40,23 @@ export class AngularFireAuthGuard implements CanActivate {
     return this.authState.pipe(
       take(1),
       authPipeFactory(next, state),
-      map(can => typeof can == "boolean" ? can : this.router.createUrlTree(<any[]>can))
+      map(can => typeof can === 'boolean' ? can :
+        Array.isArray(can) ? this.router.createUrlTree(can) : this.router.parseUrl(can)
```
Comment (NothingEverHappens) [inline src/auth-guard/auth-guard.ts:44]: Double ternary is kinda hard to read un understand (especially formatted this way).
Consider using a different approach, prob. just few "if" statements

Comment (NothingEverHappens) [inline src/auth-guard/auth-guard.ts:44]: Can we have a test for this new functionality?

**Comment by NothingEverHappens** [comment]
Note: those suggestions would be relevant once @jamesdaniels confirms the feasibility of this approach

**Comment by EdricChan03** [comment]
I can't seem to rebase the changes - was the file updated in a recent commit?

File: src/auth-guard/auth-guard.ts (hunk 1 above)
Comment (EdricChan03) [reply src/auth-guard/auth-guard.ts:44]: The current tests don't seem to test much... I'll see what I can do.

**Comment by NothingEverHappens** [comment]
> I can't seem to rebase the changes - was the file updated in a recent commit?
Yes, lots of style fixes across the codebase, may have to merge manually

**Comment by EdricChan03** [comment]
> > I can't seem to rebase the changes - was the file updated in a recent commit?
> > Yes, lots of style fixes across the codebase, may have to merge manually

I've just fixed the conflicts locally.

**Comment by jamesdaniels** [comment]
I'm good with this approach, UrlTree support would be great.

**Comment by EdricChan03** [comment]
> I'm good with this approach, UrlTree support would be great.

I don't see any way for the user to pass in a `UrlTree` to the auth guard?

**Pull request merged**